crates/sui-json-rpc-types/src/sui_object.rs

```diff
@@ -23,52 +24,77 @@ use sui_types::base_types::{
     TransactionDigest,
 };
-use sui_types::error::{UserInputError, UserInputResult};
+use sui_types::error::{SuiObjectResponseError, UserInputError, UserInputResult};
 use sui_types::gas_coin::GasCoin;
 use sui_types::move_package::{MovePackage, TypeOrigin, UpgradeInfo};
 use sui_types::object::{Data, MoveObject, Object, ObjectFormatOptions, ObjectRead, Owner};
 
-use crate::{Page, SuiMoveStruct, SuiMoveValue};
-
 #[derive(Serialize, Deserialize, Debug, JsonSchema, Clone, PartialEq, Eq)]
-#[serde(tag = "status", content = "details", rename = "ObjectRead")]
-pub enum SuiObjectResponse {
-    Exists(SuiObjectData),
-    NotExists(ObjectID),
-    Deleted(SuiObjectRef),
+pub struct SuiObjectResponse {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub data: Option<SuiObjectData>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub error: Option<SuiObjectResponseError>,
+}
+
+impl SuiObjectResponse {
+    pub fn new(data: Option<SuiObjectData>, error: Option<SuiObjectResponseError>) -> Self {
+        Self { data, error }
+    }
+
+    pub fn new_with_data(data: SuiObjectData) -> Self {
+        Self {
+            data: Some(data),
+            error: None,
+        }
+    }
+
+    pub fn new_with_error(error: SuiObjectResponseError) -> Self {
+        Self {
+            data: None,
+            error: Some(error),
+        }
+    }
 }
 
 impl SuiObjectResponse {
     pub fn move_object_bcs(&self) -> Option<&Vec<u8>> {
-        match self {
-            SuiObjectResponse::Exists(data) => match &data.bcs {
+        if let Some(data) = &self.data {
+            match &data.bcs {
                 Some(SuiRawData::MoveObject(obj)) => Some(&obj.bcs_bytes),
                 _ => None,
-            },
-            _ => None,
+            };
         }
+        None
     }
     pub fn owner(&self) -> Option<Owner> {
-        match self {
-            SuiObjectResponse::Exists(e) => e.owner,
-            SuiObjectResponse::NotExists(_) => None,
-            SuiObjectResponse::Deleted(_) => None,
+        if let Some(data) = &self.data {
+            return data.owner;
         }
+        None
     }
 
-    pub fn object_id(&self) -> ObjectID {
-        match self {
-            SuiObjectResponse::Exists(e) => e.object_id,
-            SuiObjectResponse::NotExists(ne) => *ne,
-            SuiObjectResponse::Deleted(d) => d.object_id,
+    pub fn object_id(&self) -> Option<ObjectID> {
+        match (&self.data, &self.error) {
+            (Some(obj_data), None) => Some(obj_data.object_id),
+            (None, Some(SuiObjectResponseError::NotExists { object_id })) => Some(*object_id),
+            (
+                None,
+                Some(SuiObjectResponseError::Deleted {
+                    object_id,
+                    version: _,
+                    digest: _,
+                }),
+            ) => Some(*object_id),
+            // Should never reach here.
```

Comment:
This is probably the most questionable thing here @666lcz. Do you have any suggestion?